File: 1438/solution.ts
function longestSubarray(nums: number[], limit: number): number {
  const dequeMax: [number, number][] = [];
  const dequeMin: [number, number][] = [];
  let res = 0;
  let l = 0;

  for (let r = 0; r < nums.length; r++) {
    while (dequeMax.length && nums[r] > dequeMax[dequeMax.length - 1][1]) {
      dequeMax.pop();
    }
    dequeMax.push([r, nums[r]]);

    while (dequeMin.length && nums[r] < dequeMin[dequeMin.length - 1][1]) {
      dequeMin.pop();
    }
    dequeMin.push([r, nums[r]]);

    while (dequeMax[0][1] - dequeMin[0][1] > limit) {
      l++;
      if (dequeMax[0][0] < l) {
        dequeMax.shift();
      }
      if (dequeMin[0][0] < l) {
        dequeMin.shift();
      }
    }

    res = Math.max(res, r - l + 1);
  }

  return res;
}

File: 1438/solution.py
from collections import deque


class Solution:
    def longestSubarray(self, nums: list[int], limit: int) -> int:
        dec = deque()
        inc = deque()
        l = 0
        res = 0
        for r in range(len(nums)):
            while dec and nums[r] > dec[-1][1]:
                dec.pop()
            while inc and nums[r] < inc[-1][1]:
                inc.pop()
            dec.append([r, nums[r]])
            inc.append([r, nums[r]])
            while dec[0][1] - inc[0][1] > limit:
                l += 1
                if dec[0][0] < l:
                    dec.popleft()
                if inc[0][0] < l:
                    inc.popleft()
            res = max(res, r - l + 1)
        return res
